Automattic/jetpack
# Add pre-connection partner coupon JITM

### Prerequisite

* https://github.com/Automattic/jetpack/pull/21813 - This PR depends on the coupon logic introduced here.

### Changes proposed in this Pull Request:


* Add a "Partner Pre-connection JITM" that is intended to inform non-connected Jetpack users that they have a "Hosting offer" waiting for them.

The pre-connection JITM system works in a queue, so we simply append the "Coupon Partner"  message in front and if it has been closed, the next JITM will take its place.

### Jetpack product discussion



* PT: pbtFPC-2c3-p2
* Design: p1HpG7-dvB-p2

### Does this pull request change what data or activity we track or use?



* No. This PR only touches "pre connection code" which automatically means: _no tracking_.

### Testing instructions:






* Spin up local version of this branch.
* Add `define( 'SCRIPT_DEBUG', true );` to `wp-config.php`. This is required for the JITM script to work since the minified versions doesn't exist before building the actual plugin.
* Make sure Jetpack is installed (but not connected).
* Add the _**Test partner snippet**_ snippet below to your WordPress installation (e.g.: in 0-sandbox.php if you use Jetpack Docker). This will allow the test coupon to be recognised as a valid coupon.
* Go to `/wp-admin/admin.php?jetpack-partner-coupon=JPTST_JPTA_123AB` (but don't connect). This will just store the coupon so we trigger the pre-connection JITM.
* Go to `/wp-admin/upload.php`.
* You should now see the Coupon JITM (see "Pre-connection partner coupon JITM" screenshot below).
  * If you click the "Redeem Coupon" button, you should be taken to the connection screen that asks up you setup Jetpack and claim a partner product.
  * If you close the JITM (press "X") and refresh the page, you should see another JITM. On the uploads page it's a promotion for lightning fast images (see "Other JITM" screenshot below).

_**Test partner snippet**_
```
add_filter( 'jetpack_partner_coupon_supported_partners', function ( $partners ) {
	$partners['JPTST'] = 'Jetpack Test';

	return $partners;
} );

add_filter( 'jetpack_partner_coupon_supported_presets', function ( $presets ) {
	$presets['JPTA'] = 'jetpack_backup_daily';

	return $presets;
} );
```

### Screenshots

**Pre-connection partner coupon JITM**

![Screenshot 2021-11-23 at 18 20 40](https://user-images.githubusercontent.com/3846700/143073092-b7b75e08-ab3e-4c6d-a857-8297c0c2dc88.png)

**Other JITM**

![Screenshot 2021-11-23 at 18 15 28](https://user-images.githubusercontent.com/3846700/143072556-c0776222-4995-42de-ac48-cb068c635beb.png)
